### ISSUE TITLE
Add video latency constrainable property

### ISSUE THREAD

**Issue by kuddai** [w3c/webrtc-pc#2109]
Can we extended existing [latency](https://w3c.github.io/webrtc-pc/#mediatracksupportedconstraints-mediatrackcapabilities-mediatrackconstraints-and-mediatracksettings) constrainable property API to support video MediaStreamTrack sourced by an RTCPeerConnection? This extensions would open a room for quality improvements for video MediaStreamTrack sourced by an RTCPeerConnection as you will have additional time until the video playout. For example, in the case of network issues you will have more video samples to play and recover. Or if the packets are lost then you will have more time for packets retransmission. Another problem it might solve is audio video desynchronization. Currently you have option to only increase latency for the audio track and if it has associated video track then they would be desynced.

**Comment by kuddai** [comment]
In corresponding [pull request](https://github.com/w3c/webrtc-pc/pull/2110) there was a question about whether our intention is about read only property or the actual ability to constrain latency.
We want the ability to constrain latency property.

**Comment by jan-ivar** [comment]
@kuddai thanks for clarifying! Unfortunately, constrainable properties on remote tracks is severely under-specified, with only a [single paragraph](https://w3c.github.io/webrtc-pc/#mediatracksupportedconstraints-mediatrackcapabilities-mediatrackconstraints-and-mediatracksettings).

That paragraph only really talks about "`MediaTrackSettings`" being "populated with members". I find it hard to interpret more from that than "there should be observable read-only settings". There's no requirement or mention of what should happen when e.g. *applyConstraints()* is called.

Now, Media Capture's ["the model"](https://w3c.github.io/mediacapture-main/getusermedia.html#the-model-sources-sinks-constraints-and-settings) section does imagine how this *could* work. But that section is non-normative, imaginative (automatic renegotiation from *applyConstraints*??), and somewhat outdated. It's even missing some key illustrations on this very topic. As I recall, our appetite for *action-at-a-distance* waned over the years.

In other words, an implementation today that always rejects `receiver.track.applyConstraints()` with `OverconstrainedError`, like Firefox does, would likely be considered compliant, I think.

Making this section stronger would likely require new consensus, and probably not start with *latency*.

That said, I'd probably take a PR that removes *latency* from the list of exclusions.

> Currently you have option to only increase latency for the audio track

Where do you see that?

If this is important to you, I would start with collecting support for the use cases and problem first, and then look for the right API second: Is the best place for this control knob really on every track clone, or the receiver, the stream (which is a synchronization group after all) or even the pc itself?

**Comment by kuddai** [comment]
It can be seen [here](https://www.w3.org/TR/mediacapture-streams/#def-constraint-latency), if you scroll up to the header of the table you will see "The following constrainable properties are defined to apply **only to audio** MediaStreamTrack objects"

**Comment by kuddai** [comment]
As for the use case I can share how we want to use it.
In the real time communication if you have more than two participants then quite often some of the participants doesn't speak. During this time we want to increase latency on MediaStreamTracks sourced by an RTCPeerConnection. Increased delay means, depending on your implementation of course, that you buffer can accumulate more packets, and you have more media to play in case of problems which results in much better media quality. Increased delay is undesirable during conversation but that is why we want to increase only for participants who don't speak, for example muted their microphone.

**Comment by kuddai** [comment]
And to make it explicit, this one is high priority for me.

**Comment by guidou** [comment]
Separate from this proposal, the spec language IMO does not forbid applyConstraints() from working on tracks sourced by a peer connection.  It just says "The basics of ... is outlined in [GETUSERMEDIA]".

In Chrome in particular, applyConstraints() allows changing resolution for peerconnection tracks  via rescaling.

With regards to this proposal, I think giving a different meaning to the latency property in remote tracks, as proposed in the #2110, (instead of saying it doesn't exist) should neither mandate nor forbid constraining its value.

**Comment by jan-ivar** [comment]
> the spec language IMO does not forbid ...

Most behavior is not forbidden by specs. My reading of specs is that whatever they say to do is standard; whatever they don't is not.

> In Chrome in particular, applyConstraints() allows changing resolution for peerconnection tracks via rescaling.

If we want this behavior to be web compatible, I suggest we specify it. That's [what we did](https://w3c.github.io/mediacapture-screen-share/#constrainable-properties-for-captured-display-surfaces) for screen-capture.

I'm happy to remove the *"will always be missing"*, to avoid implying we will never do this.

Beyond that, this looks like a new feature in 1.0. which I think we should resist at this point, and encourage extension specs for.

**Comment by guidou** [comment]
I interpret the text "The basics of MediaTrackSupportedConstraints, MediaTrackCapabilites, MediaTrackConstraints and MediaTrackSettings is outlined [GETUSERMEDIA]" to mean that applyConstraints(), getCapabilities() and getSettings() should work for PeerConnection tracks in the way that is indicated in the getUserMedia() spec.  Thus, I do not see why these functions should not work for PeerConnection tracks.

I do not interpret the text saying that the only properties visible in MediaTrackSettings are those actually available on a PeerConnection to mean that getSettings() is the only function from the constraintable pattern that should work on PeerConnection tracks. If that is the way it should be interpreted, I think it needs a rewrite because that is not what is says.

**Comment by jan-ivar** [comment]
It doesn't specify *how* it should work. Implementors should not have to infer this. I think there's precedent here in https://github.com/w3c/mediacapture-screen-share/issues/31.

*MediaTrackSupportedConstraints* is returned from [navigator.mediaDevices.getSupportedConstraints()](https://w3c.github.io/mediacapture-main/getusermedia.html#dom-mediadevices-getsupportedconstraints). How exactly is it relevant to RTCPeerConnection? 

Should the following blank the remote track and  fire `overconstrainederror` and ?
```js
await transceiver.receiver.track.applyConstraints({width: {min: 1280}}); // succeeds
transceiver.receiver.track.onoverconstrainederror = () => { /* should this fire? */ };
// Network conditions or other side using replaceTrack causes lower resolution to appear
```

**Comment by guidou** [comment]
> It doesn't specify _how_ it should work. Implementors should not have to infer this. I think there's precedent here in [w3c/mediacapture-screen-share#31](https://github.com/w3c/mediacapture-screen-share/issues/31).

The getUserMedia spec specifies pretty well how it should work. If it shouldn't work as explained in the getUserMedia spec, then the whole 9.3.1 section should be removed or rewritten.

> 
> _MediaTrackSupportedConstraints_ is returned from [navigator.mediaDevices.getSupportedConstraints()](https://w3c.github.io/mediacapture-main/getusermedia.html#dom-mediadevices-getsupportedconstraints). How exactly is it relevant to RTCPeerConnection?
> 
> Should the following blank the remote track and fire `overconstrainederror` and ?
> 
> ```js
> await transceiver.receiver.track.applyConstraints({width: {min: 1280}}); // succeeds
> transceiver.receiver.track.onoverconstrainederror = () => { /* should this fire? */ };
> // Network conditions or other side using replaceTrack causes lower resolution to appear
> ```

Yes, it should fire. And, BTW, Chrome is implementing that behavior.

**Comment by jan-ivar** [comment]
> Yes, it should fire. And, BTW, Chrome is implementing that behavior.

@guidou When I [try it](https://jsfiddle.net/jib1/x7r4pknu/) in Chrome I get `applyConstraints not supported for this track`:
```
getSettings: {"deviceId":"b88716e4-4d43-45d3-8a96-6b6748a6a7ad","resizeMode":"none"}
getCapabilities: {"deviceId":"b88716e4-4d43-45d3-8a96-6b6748a6a7ad","facingMode":[],"resizeMode":["none","crop-and-scale"]}
OverconstrainedError: applyConstraints not supported for this track
```
I get similar output in Canary except it crashes. Is it behind a pref?

FWIW, I would not expect a peer connection to have a `deviceId`, nor mention capabilities that do not make sense, like `facingMode`. [resizeMode](https://w3c.github.io/mediacapture-main/getusermedia.html#def-constraint-resizeMode) mentions camera. A peer connection is not a camera.

Basically, this is not what I expected to see, so I think we need to specify this. I've opened https://github.com/w3c/webrtc-pc/issues/2121. Let's discuss there, and leave this issue to discuss the merits of video latency for remote tracks.

**Comment by guidou** [comment]
applyConstraints() on remote tracks landed recently, so you need a very recent canaYour example works for me on a recent canary (although getSettings

WRT to resizeMode, I would argue that the spec should say source instead of camera.  The definition of what the property does applies to any source that provides frames, and even getUserMedia() can support devices that are not cameras (but use the same system interfaces typically used by webcams).

**Comment by jan-ivar** [comment]
@guidou Thanks for the info. Looking forward to trying it! FWIW the crash I got was on 74.0.3726.0 on Mojave.

**Comment by guidou** [comment]
@jan-ivar Thanks for the info.
Reporting the deviceId is indeed a bug. In Chrome PC tracks have an internal fake deviceID and the code is exposing like for gUM tracks. We're tracking that bug in https://crbug.com/938840.

As for the crash, thanks for the report. I just tried it on Mac and also got the crash. It works as expected on Linux for me. Filed https://crbug.com/939682

**Comment by fippo** [comment]
what is the rationale for making this controllable via the track and not the rtpReceiver? (other than the receiver not having a setParameters)

**Comment by kuddai** [comment]
> what is the rationale for making this controllable via the track and not the rtpReceiver? (other than the receiver not having a setParameters)

Mostly, because the latency parameter we want to change is quite similar to the one which has been already described by MediaStreamTrack applyConstraints algorithm. On the rtpReceiver even getParameters are not having anything similar.

**Comment by jan-ivar** [comment]
> (other than the receiver not having a setParameters)

There's no rule saying we can't specify a modifiable `receiver.latency` attribute.

Symmetry spell broken. 🧙

**Comment by kuddai** [comment]
Closed in favor of #2139

**Comment by kuddai** [comment]
Closed in favor of #2139
